server/controllers/banner.js

```diff
@@ -1,115 +1,150 @@
+                const now = moment();
 
-            let allowedBanners = availableBanners
-                .filter(banner => !banner.isDefault)
-                .filter(banner => 
-                    moment(banner.begin).isBefore(now)
-                    && moment(banner.end).isAfter(now)
-                )
-            ;
+                const availableBanners = this.getValues(banners)
+                    .filter(this.containsCompany(params.company))
+                ;
 
-            if(allowedBanners.length === 0) {
-                allowedBanners = availableBanners
-                    .filter(banner => banner.isDefault)
-            }
+                let allowedBanners = availableBanners
+                    .filter(banner => !banner.isDefault)
+                    .filter(banner =>
+                        moment(banner.begin).isBefore(now)
+                        && moment(banner.end).isAfter(now)
+                    )
+                ;
 
-            const randomBanner = allowedBanners[this.rand(0, allowedBanners.length -1)];
+                if(allowedBanners.length === 0) {
+                    allowedBanners = availableBanners
+                        .filter(banner => banner.isDefault)
+                }
 
-            resolve(randomBanner.path);
+                const randomBanner = allowedBanners[this.rand(0, allowedBanners.length -1)];
+
+                resolve(this.getPictureUrlFromS3Path(randomBanner.path));
+            });
         });
     }
 
     containsCompany(companyName) {
+
         return (banner) => {
+
             if (companyName) {
                 return banner.companies[companyName];
             }
+
             return true;
         }
     }
 
     uploadBanner(bannerFolder, fieldname, file, filename) {
 
-        return new Promise((resolve, reject) => {
-
-            mkdirp(bannerFolder, (err) => {
-
-                if(err) console.error(err);
+        const fsImpl = this.getFs();
 
-                console.log("Uploading: " + filename);
+        return new Promise((resolve, reject) => {
 
-                const [_, name, extension] = /(.*)\.(.*)/.exec(filename);
-                const newFilename = `${Date.now()}.${extension}`;
-                const fstream = fs.createWriteStream(bannerFolder + newFilename);
+            console.log("Uploading: " + filename);
 
-                file.pipe(fstream);
+            const [_, name, extension] = /(.*)\.(.*)/.exec(filename);
+            const newFilename = `${Date.now()}.${extension}`;
+            const fstream = fsImpl.createWriteStream(bannerFolder + newFilename);
 
-                fstream.on("close", () => {
-                    console.log("Upload succeed !");
-                    resolve(newFilename);
-                });
+            file.pipe(fstream);
+            fstream.on("close", () => {
+                console.log("Upload succeed !");
+                resolve(newFilename);
             });
+
+            resolve(newFilename);
```

Comment:
deux resolves ? C'est normal ?

Reply:
Fixed in 5d89a4203d9a330ac94b9e99d0261879b0d7bbc5. S3FS doesn't implements on close stream function.